Architecture_Engine/Query/Distance.cs

```diff
@@ -0,0 +1,90 @@
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Architecture.BuildersWork;
+using BH.oM.Architecture.Elements;
+using BH.oM.Geometry;
+using System.Collections.Generic;
+using System.Linq;
+using BH.oM.Spatial.ShapeProfiles;
+using BH.Engine.Geometry;
+using BH.oM.Geometry.CoordinateSystem;
+
+namespace BH.Engine.Architecture
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static double Distance(this Opening opening1, Opening opening2, double maxExpectedDistance = double.NaN, double distanceTolerance = Tolerance.Distance, double angleTolerance = Tolerance.Angle)
+        {
+
+            if (opening1 == null || opening2 == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Opening is null");
+                return double.NaN;
+            }
+
+            double minDistance = double.MaxValue;
+
+            Point centerPoint2 = opening2.CoordinateSystem.Origin;
+            Plane plane2 = new Plane { Origin = centerPoint2, Normal = opening2.CoordinateSystem.Z };
+
+            if (!opening1.IsValidForDistanceCheck(opening2, maxExpectedDistance, distanceTolerance, angleTolerance)) 
+            {
+                BH.Engine.Reflection.Compute.RecordError("Distance can't be calculated");
+                return double.NaN;
+            }
+
+            else
+            {               
+                foreach (ICurve edge1 in opening1.Profile.Edges)
+                {
+                    foreach (ICurve edge2 in opening2.Profile.Edges)
```

Comment:
This is very inefficient because you need to extract the edges of `opening2` and orient them for every edge of `opening1`. Wouldn't it be better to it once before entering any `for` loop?

```
List<ICurve> edges2 = opening2.Profile.Edges.Select(x => x.Orient(global, opening2.CoordinateSystem).IProject(plane1)).ToList();
foreach (ICurve edge1 in opening1.Profile.Edges)
{
    //do your checks here
    //could also orient edges of opening1 before entering for loop for better code clarity
}
```